### DISCUSSION TITLE
Catching exceptions across all route handlers

### DISCUSSION THREAD

**Discussion by nomadeon** [dotnet/WatsonWebserver#149] [Q&A] [answered]
Is there a way to catch exceptions across all async route handlers (without putting a try/catch in each)?

I've tried wiring up to `TaskScheduler.UnobservedTaskException` and `AppDomain.CurrentDomain.UnhandledException` with no luck. I'm aware of the issues with async void and that is not the method signature I'm using for the route:
`async Task PositionsRoute(HttpContextBase ctx)`

**Comment by jchristn** [comment]
Hi @nomadeon at this point there is not; there are default pages that can be configured by status, but I'm not sure that will meet your needs.  Would you prefer something that allowed you to attach an exception event handler to every subordinate route within the routing group?  Something like that is feasible.

**Reply by nomadeon** [reply]
> Hi @nomadeon at this point there is not; there are default pages that can be configured by status, but I'm not sure that will meet your needs. Would you prefer something that allowed you to attach an exception event handler to every subordinate route within the routing group? Something like that is feasible.

Yes, that would do nicely. Perhaps something like:
`_server.Routes.PreAuthentication.OnException(ExceptionHandler);`
Where `ExceptionHandler` is simply an `Action<Exception,HttpRequestBase>`

It would be fine if the server still returned 500 with no way to play with that in the exception handler. Just need it for logging/debugging/reporting.

EDIT: would like to have some info about request available in the exception handler. I've put in HttpRequestBase as a suggestion.

**Reply by jchristn** [reply, marked as answer]
Hi @nomadeon thanks for the suggestion, just implemented this in NuGet v6.2.0.  Please let me know what you think.

**Reply by nomadeon** [reply]
Most excellent!
